src/dune_pkg/rev_store.ml

```diff
@@ -391,22 +391,88 @@ module Entry = struct
   ;;
 end
 
-module Submodule = struct
-  (* a submodule in [.gitmodules] can also have a [branch] but given we only
-     need to resolve the commit object, we don't have to care about the
-     tracking branch *)
+module At_rev = struct
+  type repo = t
+
   type t =
-    { path : Path.Local.t
+    { repo : repo
+    ; revision : Rev.t
     ; source : string
+    ; files : File.Set.t
     }
 
-  let parse lines =
-    match Git_config_parser.parse lines with
-    | Error err ->
-      (* CR-rgrinberg: the loc needs to be pulled from the git URL *)
-      User_error.raise [ Pp.textf "Failed to parse submodules: %s" err ]
-    | Ok cfg ->
-      List.filter_map cfg ~f:(fun { name; arg = _; bindings } ->
+  module Config = struct
+    type bindings = string * string
+
+    type section =
+      { name : string
+      ; arg : string option
+      ; bindings : bindings list
+      }
+
+    type t = section list
+
+    module KV = struct
+      module T = struct
+        type t = string * string option
+
+        let compare = Tuple.T2.compare String.compare (Option.compare String.compare)
+        let to_dyn = Tuple.T2.to_dyn Dyn.string (Dyn.option Dyn.string)
+      end
+
+      include Comparable.Make (T)
+    end
+
+    let config { dir } revision path : t Fiber.t =
+      let (Rev.Rev rev) = revision in
+      let location = [ "--blob"; sprintf "%s:%s" rev (Path.Local.to_string path) ] in
+      let command = [ "config"; "--list" ] @ location in
```

Comment:
I don't have the big picture, but instead of reading all the config at once, we could just read each key as needed. Especially if we're going to `--set` things, this could avoid some race conditions if the config file changes.

Reply:
The problem with this is that for each submodule we'd need to call `git config` at least twice (for `submodule.<name>.path` and `submodule.<name>.url`) and to get the list of `<name>` submodules we have to call `git config --list` anyway, at which point we get all the parsed data.

I don't think this is a big issue, since we probably can't `--set` on `--blob` anyway and the `.gitmodule` files that we parse are usually fairly short.